From fea2e762b6c740827caf67ad4d9d49ad869de148 Mon Sep 17 00:00:00 2001
From: Mark Thomas <markt@apache.org>
Date: Mon, 3 Feb 2014 21:55:38 +0000
Subject: [PATCH] Fix DBCP-383 Avoid NullPointerException and throw an
 XAException if an attempt is made to commit the current transaction for a
 connection when no transaction has been started.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/dbcp/branches/DBCP_1_5_x_BRANCH@1564086 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                              |  5 +++++
 .../dbcp/managed/LocalXAConnectionFactory.java       | 12 ++++++++++--
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 019e386c26..ac0ab0893a 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -94,6 +94,11 @@ The <action> type attribute can be add,update,fix,remove.
       <action dev="markt" issue="DBCP-398" type="fix">
         Clarify Jaavdoc for isClosed() method of PoolableConnection.
       </action>
+      <action dev="markt" issue="DBCP-383" type="fix">
+        Avoid NullPointerException and throw an XAException if an attempt is
+        made to commit the current transaction for a connection when no
+        transaction has been started.
+      </action>
     </release>
     <release version="1.4.1" date="TBD" description="TBD">
       <action dev="psteitz" issue="DBCP-367" type="fix" due-to="Ken Tatsushita">
diff --git a/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java b/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java
index a68996ab6f..efb2e02056 100644
--- a/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java
+++ b/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java
@@ -197,8 +197,16 @@ public synchronized int prepare(Xid xid) {
          * @throws XAException if connection.commit() throws a SQLException
          */
         public synchronized void commit(Xid xid, boolean flag) throws XAException {
-            if (xid == null) throw new NullPointerException("xid is null");
-            if (!this.currentXid.equals(xid)) throw new XAException("Invalid Xid: expected " + this.currentXid + ", but was " + xid);
+            if (xid == null) {
+                throw new NullPointerException("xid is null");
+            }
+            if (this.currentXid == null) {
+                throw new XAException("There is no current transaction");
+            }
+            if (!this.currentXid.equals(xid)) {
+                throw new XAException("Invalid Xid: expected " +
+                        this.currentXid + ", but was " + xid);
+            }
 
             try {
                 // make sure the connection isn't already closed
